### DISCUSSION TITLE
Q: how to restrict a rule

### DISCUSSION THREAD

**Discussion by cppden** [taocpp/PEGTL#45] [Q&A] [answered]
Is there a way to restrict matching in the following case:
```cpp
struct param_a : seq< istring<'a','='>, digit > {};
struct param_b : seq< istring<'b','='>, alpha > {};
struct param_other : seq< alpha, one<'='>, alnum > {};
struct params : sor< param_a, param_b, param_other > {};
```
Namely, I don't want grammar to accept neither 'a=c' nor 'b=3' but the last choice allows it. Please note that this is simplified example and in reality there's about 10 choices with lengthy name part (instead of 'a' and 'b'). Therefore it's really hard and clumsy to exclude in param_other case all previous cases. I wonder if there's more elegant way to restrict the rule?
Thanks a lot!

**Comment by ColinH** [comment]
What we usually do in this situation is to use `if_must<>` instead of `seq<>` in `param_a` and `param_b`. (That way there will be no backtracking once you get past the "distinguishing" part of the `sor<>`-branch.) If, as per the other question, you are trying to eliminate the use of `must<>`, you could use the new `minus<>` rule to exclude the first cases from `param_other`, but that is slow and ugly. (Probably very much more slow and ugly than using `must<>` and catching the exceptions.)

**Comment by d-frey** [comment, marked as answer]
You could use nested `if_then_else` rules, e.g.

    if_then_else< istring<'a','='>, digit,
    if_then_else< istring<'b','='>, alpha,
    param_other > >

which is strangely formatted when it comes to nesting and which leaves the question of where to attach the actions. But from a pure parsing POV, it will do exactly what you need.

I think we could also easily build a better interface, maybe a "sequenced switch" with an interface like

    sswitch< istring<'a','='>, digit,
             istring<'b','='>, alpha,
             param_other >

and maybe allow specializing on `sswitch_case<istring<'a','='>>` for the actions and passing the whole matching input of the case. @ColinH What do you think?

**Comment by cppden** [comment]
Thanks a lot! If-then-else looks fine to me, and I hope actions can be attached to digit and alpha since their values are all what's needed.
Regarding sequenced switch it seems less readable/explicit at 1st glance... And personally I prefer minimal interface. :) More alternatives, more fuss.
Thank you very much for the help and great lib! 
And pardon me for responding much slower than you reply to my questions. ;)

**Comment by d-frey** [comment]
You could attach actions to `digit` or `alpha`, but that will most probably lead to a lot of problems as the rest of your grammar might also contain those (even `param_other` in your example). The solution is, as usual, to create anchor points explicitly:

```cpp
struct intro_a : istring<'a','='> {};
struct value_a : digit {};

struct intro_b : istring<'b','='> {};
struct value_b : alpha {};

struct other : seq< alpha, one<'='>, alnum > {};

struct params : if_then_else< intro_a, value_a,
                if_then_else< intro_b, value_b,
                other > > {};
```

Then just attach actions to `value_a`, etc.

Also: Phew! Good to know I'm not the only one who prefers minimalistic interfaces :)
